cli/cmd/cluster/certificate-rotator.go

```diff
@@ -0,0 +1,220 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+	"github.com/kinvolk/lokomotive/pkg/platform"
+)
+
+type certificateRotatorConfig struct {
+	clientSet            *kubernetes.Clientset
+	newCACert            string
+	logger               *log.Entry
+	daemonSetsToRestart  []platform.Workload
+	deploymentsToRestart []platform.Workload
+}
+
+type certificateRotator struct {
+	config certificateRotatorConfig
+}
+
+// CertificateRotateOptions contains the options for the RotateCertificates function.
+type CertificateRotateOptions struct {
+	Confirm    bool
+	Verbose    bool
+	ConfigPath string
+	ValuesPath string
+}
+
+// RotateCertificates replaces all certificates in a cluster.
+func RotateCertificates(contextLogger *log.Entry, options CertificateRotateOptions) error {
+	cc := clusterConfig{
+		verbose:    options.Verbose,
+		configPath: options.ConfigPath,
+		valuesPath: options.ValuesPath,
+	}
+
+	c, err := cc.initialize(contextLogger)
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	if err := clusterRotatable(c); err != nil {
+		return fmt.Errorf("cannot rotate cluster: %w", err)
+	}
+
+	// Tainting certificates so they get rotated.
+	if err := c.taintCertificates(); err != nil {
+		return fmt.Errorf("tainting certificate resources: %w", err)
+	}
+
+	// Apply the Terraform changes to replace tainted resources.
+	if err := c.platform.Apply(&c.terraformExecutor); err != nil {
+		return fmt.Errorf("applying platform: %v", err)
+	}
+
+	return rotateControlPlateCerts(contextLogger, cc)
+}
+
+func clusterRotatable(c *cluster) error {
+	exists, err := clusterExists(c.terraformExecutor)
+	if err != nil {
+		return fmt.Errorf("checking if cluster exists: %w", err)
+	}
+	if !exists {
+		return errors.New("cannot rotate certificates on a non-existent cluster")
+	}
+
+	if c.platform.Meta().Managed {
+		// TODO: do we want to error here?
+		return errors.New("cannot rotate certificates on a managed cluster")
+	}
+
+	return nil
+}
+
+func rotateControlPlateCerts(contextLogger *log.Entry, cc clusterConfig) error {
+	c, err := cc.initialize(contextLogger)
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	kg := kubeconfigGetter{
+		platformRequired: true,
+		clusterConfig:    cc,
+	}
+
+	kubeconfig, err := kg.getKubeconfig(contextLogger, c.lokomotiveConfig)
+	if err != nil {
+		return fmt.Errorf("getting kubeconfig: %v", err)
+	}
+
+	contextLogger.Log(log.InfoLevel, "Applying a controlplane update with the new CA")
+
+	if err := c.upgradeControlPlane(contextLogger, kubeconfig); err != nil {
```

Comment:
>This commit is quite large

IMO config updates part should be broken down into separate commits. Actually a lot of those things can be broken down.

>I'm following correctly: the service accounts that we wait to be updated inside this function are updated because of this part? And those are updated because the sha changed, so the pods are re-created with a new service account?

Correct. Once kube-controller-manager is restarted with new CA certificate, controller will update all service accounts with new certificate. Finally, we restart all controlplane components to ensure they picked up new CA (this part is in theory not needed, as client-go should be able to pick them up on the fly and secrets are also updated on the fly, but I was not able to get it to work).